Haptics/app/src/main/java/com/example/android/haptics/samples/ui/bounce/BounceRoute.kt

```diff
@@ -0,0 +1,327 @@
+import androidx.compose.material.Icon
+import androidx.compose.material.MaterialTheme
+import androidx.compose.material.Text
+import androidx.compose.material.icons.Icons
+import androidx.compose.material.icons.outlined.TouchApp
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.setValue
+import androidx.compose.ui.Alignment
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.draw.clip
+import androidx.compose.ui.platform.LocalContext
+import androidx.compose.ui.res.stringResource
+import androidx.compose.ui.tooling.preview.Preview
+import androidx.compose.ui.unit.Dp
+import androidx.compose.ui.unit.dp
+import com.example.android.haptics.samples.R
+import com.example.android.haptics.samples.ui.components.Screen
+import com.example.android.haptics.samples.ui.modifiers.noRippleClickable
+import com.example.android.haptics.samples.ui.theme.HapticSamplerTheme
+import kotlin.math.pow
+
+/**
+ * The two states of the animated ball. Start refers to the elevated state in which
+ * the ball drops from, and End once the ball is at rest on the "floor".
+ */
+private enum class BallState {
+    Start,
+    End,
+}
+
+private val BALL_SIZE = 64.dp
+private val BALL_DROP_HEIGHT_DP = 300.dp
+private val BALL_START_POSITION = -(BALL_DROP_HEIGHT_DP)
+private val BALL_END_POSITION = 0.dp
+// If recomposition occurs within these ranges, we consider it a collision and will vibrate.
+// Speed of animation impacts the ranges we select. For example, we must have a larger range to
+// detect initial collision with floor as ball animates quickly through it, otherwise
+// it'd be possible for no recomposition to occur within range.
+private val DISTANCE_FROM_END_POSITION_FOR_COLLISION = 5.dp
+private val DISTANCE_FROM_START_POSITION_FOR_COLLISION = 1.dp
+
+private val FLOOR_SIZE = 80.dp
+
+private const val DROP_ANIMATION_TIME_MS = 3000
+private const val RESET_ANIMATION_TIME_MS = 1000
+
+@Composable
+fun BounceRoute(viewModel: BounceViewModel) {
+    BounceExampleScreen(messageToUser = viewModel.messageToUser)
+}
+
+@Composable
+private fun BounceExampleScreen(messageToUser: String) {
+    var ballState by remember { mutableStateOf(BallState.Start) }
+    var bounceCount by remember { mutableStateOf(0) }
+
+    val vibrator = LocalContext.current.getSystemService(Context.VIBRATOR_SERVICE) as Vibrator
+
+    var transitionData = updateTransitionData(
+        ballState = ballState
+    )
+    val collisionData = updateCollisionData(transitionData)
+
+    // Execute a click vibration once the ball has been reset.
+    var hasVibratedForReset by remember { mutableStateOf(false) }
+    if (collisionData.collisionWithReset) {
+        if (!hasVibratedForReset) {
+            clickVibration(vibrator)
+            hasVibratedForReset = true
+        }
+    }
```

Comment:
Can we reset hasVibratedForReset here in a else clause, like we do below for hasVibratedForBallContact, to keep this flag controlled by at a single place?

Reply:
Good call. Makes sense.